### PR TITLE
Add poetry2nix support: apps and dev env

### PR DESCRIPTION
**Pull request by jbgi**
Now `nix develop` has all `tool.poetry.scripts` in `PATH`.

Also `nix develop .#dev` can now be used to get an environment with poetry and all pythons dependencies (other needed dependencies should probably be added to it).

### CODE REVIEW

File: flake.nix [hunk 1]
```diff
@@ -17,22 +20,67 @@
...
         let
           pkgs = nixpkgs.legacyPackages.${system};
-          python3 = pkgs.python3.override {
-            packageOverrides = self: _: {
-              allure = self.callPackage ./nix/allure.nix { };
-              pytest-allure = self.callPackage ./nix/pytest-allure.nix { };
-              pytest-select = self.callPackage ./nix/pytest-select.nix {};
-              cardano-clusterlib = self.callPackage ./nix/cardano-clusterlib.nix { };
-            };
+          p2n = poetry2nix.legacyPackages.${system};
```
Comment (angerman) [inline flake.nix:28]:
```suggestion
          # As we are using [poetry](...) as the python dependency manager for cardano-node-tests, we will be using
          # [poetry2nix](https://github.com/nix-community/poetry2nix) to convert the poetry project (pyproject.toml,
          # and poetry.lock) into a nix-buildable expression. This is preferable over using `pkgs.python3.withPackages`
          # as it adheres to the poetry setup instead of replicating it in nix again.
          
          p2n = poetry2nix.legacyPackages.${system};
```

File: flake.nix [hunk 2]
```diff
@@ -17,22 +20,67 @@
...
+          # Workarounds `ModuleNotFoundError: No module named 'poetry'`
+          # cf. https://github.com/nix-community/poetry2nix/blob/master/docs/edgecases.md#modulenotfounderror-no-module-named-packagename
+          pypkgs-build-requirements = {
+            pytest-metadata = [ "poetry" ];
+            pytest-select = [ "poetry" ];
+            m2r2 = [ "setuptools" ];
+            sphinxemoji = [ "setuptools" ];
+            sphinxcontrib-jquery = [ "flit-core" ];
+          };
+          p2n-overrides = p2n.defaultPoetryOverrides.extend (self: super:
```
Comment (angerman) [inline flake.nix:38]:
```suggestion
          # for the time being we need some minor adjustments to the constructed poetry project
          # in nix.
          #
          # 1. poetry2nix fails to see the setuptools-scm dependency, hence we need to inject it by hand into Cardano-cluster lib.
          # 2. Because transitively depend on `py`, (through `pytest-html`), we need to drop the module from `pytest`.
          # 
          p2n-overrides = p2n.defaultPoetryOverrides.extend (self: super:
```

File: flake.nix [hunk 3]
```diff
@@ -17,22 +20,67 @@
...
+            builtins.mapAttrs
```
Comment (angerman) [inline flake.nix:39]: I find this part really hard to parse. What are we trying to do here?

File: flake.nix [hunk 4]
```diff
@@ -17,22 +20,67 @@
...
+              (package: build-requirements:
+                (builtins.getAttr package super).overridePythonAttrs (old: {
+                  buildInputs = (old.buildInputs or [ ]) ++ (builtins.map (pkg: if builtins.isString pkg then builtins.getAttr pkg super else pkg) build-requirements);
```
Comment (angerman) [inline flake.nix:42]: Could this be improved by selectively using `with builtins;`?
```suggestion
                  buildInputs = (old.buildInputs or [ ])
                             ++ (with builtins; map (pkg: if isString pkg then getAttr pkg super else pkg) build-requirements);
```
isn't `getAttr pkg super` identical to `super.${pkg}`? Which _I_ find more readable, maybe I'm an outlier.
So this would be `(with builtins; map (pkg: if isString pkg then super.${pkg} else pkg)`.

File: flake.nix (hunk 3 above)
Comment (angerman) [reply flake.nix:39]: Maybe something like:
```
# For each `package`, and its `build-requirements`, add the `build-requirements` to the `buildInputs`. In the case # where a `build-requirement` is referred to by name, resolve it in the `super` pkg set.
```

However, I still don't understand _why_ this is necessary?

**Review by angerman — CHANGES_REQUESTED** [review]
I think we need to add some clarity here.

File: flake.nix (hunk 3 above)
Comment (jbgi) [reply flake.nix]: Just switched to using sdist instead of building from source (except for pytest): all those overrides for injecting the build tools are not necessary anymore.

**Review by angerman — APPROVED** [review]
LGTM. Though I can't say if this properly works, as I believe we lack the necessary CI to test this?

**Comment by angerman** [comment]
@jbgi please consider adding another ticket about adding GHA's CI (e.g. install-nix-action, and just `nix develop` or something to test that the shell actually comes up).

**Review by mkoura — APPROVED** [review]
Tested, LGTM :+1: :1st_place_medal:

**Comment by mkoura** [comment]
@angerman @jbgi That's a good idea with CI for testing nix shell.
I've just added https://github.com/input-output-hk/cardano-node-tests/blob/master/.github/workflows/nix_smoke.yaml
It launches noop pytest in nix dev shell, in the same way that we launch our normal testing jobs.
